File: GuiRunner/VeiwModels/TestsMainModel.cs
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Collections.Specialized;
using System.ComponentModel;
using System.Linq;
using System.Windows.Input;
using FeldSpar.ClrInterop;
using FeldSpar.Framework;

namespace FeldSparGuiCSharp.VeiwModels
{
    public interface ITestsMainModel
    {
        bool IsRunning { get; set; }
        string Description { get; }
        ITestDetailModel Selected { get; set; }
        ObservableCollection<ITestAssemblyModel> Assemblies { get; set; }
        TestResult[] Results { get; }
        ITestDetailModel[] Tests { get; }
        ICommand RunCommand { get; }
        ICommand AddCommand { get; }
        void Run(object ignored);
        void Add(object ignored);
    }

    public class TestsMainModel : PropertyNotifyBase, ITestsMainModel
    {
        private string description;
        private ITestDetailModel selected;

        public TestsMainModel()
        {
            var itemsRemovedActions = new[] { NotifyCollectionChangedAction.Remove, NotifyCollectionChangedAction.Replace };

            Assemblies = new ObservableCollection<ITestAssemblyModel>();
            Assemblies.CollectionChanged += (sender, args) =>
            {
                var testChanged = false;
                foreach (INotifyPropertyChanged newItem in args.NewItems)
                {
                    newItem.PropertyChanged += ItemOnPropertyChanged;
                    testChanged = true;
                }

                if (testChanged)
                {
                    OnPropertyChanged("Tests");
                }

                if (itemsRemovedActions.All(x => args.Action != x))
                {
                    return;
                }

                testChanged = false;
                foreach (INotifyPropertyChanged oldItem in args.OldItems)
                {
                    oldItem.PropertyChanged -= ItemOnPropertyChanged;
                    testChanged = true;
                }

                if (testChanged)
                {
                    OnPropertyChanged("Tests");
                }
            };
        }

        private void ItemOnPropertyChanged(object sender, PropertyChangedEventArgs propertyChangedEventArgs)
        {
            if (propertyChangedEventArgs.PropertyName == "Results")
            {
                OnPropertyChanged("Results");
            }
            if (propertyChangedEventArgs.PropertyName == "Tests")
            {
                OnPropertyChanged("Tests");
            }
        }

        public bool IsRunning { get; set; }

        public void Run(object ignored)
        {
            IsRunning = true;
            foreach (var testAssemblyModel in Assemblies)
            {
                var model = testAssemblyModel;
                model.Run(ignored);
            }

            IsRunning = false;
        }

        public void Add(object ignored)
        {
            var fileOpen = new Microsoft.Win32.OpenFileDialog {Filter = "test suites|*.dll;*.exe", Multiselect = false};

            var result = fileOpen.ShowDialog();
            if (result != true)
            {
                return;
            }

            if (Assemblies.All(assembly => assembly.AssemblyPath != fileOpen.FileName))
            {
                Assemblies.Add(new TestAssemblyModel(fileOpen.FileName));
            }
        }

        public string Description
        {
            get { return description; }
            private set
            {
                if (value == description)
                {
                    return;
                }

                description = value;
                OnPropertyChanged();
            }
        }

        public ITestDetailModel Selected
        {
            get { return selected; }
            set
            {
                if (selected == value)
                {
                    return;
                }

                selected = value;
                OnPropertyChanged();

                Description = selected.FailDetail + "";
            }
        }

        public ObservableCollection<ITestAssemblyModel> Assemblies { get; set; }

        private T[] GetTestItems<T>(Func<ITestAssemblyModel, IEnumerable<T>> selector) { return Assemblies.SelectMany(selector).ToArray(); }

        public TestResult[] Results { get { return GetTestItems(assembly => assembly.Results); } }

        public ITestDetailModel[] Tests { get { return GetTestItems(assembyly => assembyly.Tests); } }

        public ICommand RunCommand { get { return new DelegateCommand(Run, _ => !IsRunning); } }

        public ICommand AddCommand { get { return new DelegateCommand(Add); } }
    }
}

File: FeldSpar.InterOpWpfEngine/TestsMainModel.fs
namespace FeldSpar.Api.Engine.ClrInterop.ViewModels
open System
open FeldSpar.Framework
open System.ComponentModel
open System.Collections.ObjectModel
open System.Collections.Generic
open System.Windows.Input
open System.Collections.Specialized

type TestsMainModel () as this =
    inherit PropertyNotifyBase()

    let mutable assemblies = new ObservableCollection<ITestAssemblyModel>()
    let mutable isRunning = false
    let mutable selected : ITestDetailModel = Defaults.emptyTestDetailModel

    let runIt runner (this:TestsMainModel) = 
        let self = this :> ITestsMainModel
        self.IsRunning <- true
        for testAssemblyModel in self.Assemblies do
            let model = testAssemblyModel
            model |> runner
            ()

        self.IsRunning <- false

    do
        let itemsRemovedActions = [ NotifyCollectionChangedAction.Remove; NotifyCollectionChangedAction.Replace ]
        assemblies.CollectionChanged.Add
            (
                fun args -> 
                    for newItem in args.NewItems do
                        let newItem = newItem :?> INotifyPropertyChanged
                        newItem.PropertyChanged.AddHandler(this.itemOnPropertyChangedHandler)

                    if args.NewItems.Count > 0 then this.OnPropertyChanged("Tests")

                    if itemsRemovedActions |> List.filter (fun action -> args.Action = action) |> List.length > 0
                    then 
                        for oldItem in args.OldItems do
                            let oldItem = oldItem :?> INotifyPropertyChanged
                            oldItem.PropertyChanged.RemoveHandler(this.itemOnPropertyChangedHandler)

                        if args.OldItems.Count > 0 then this.OnPropertyChanged("Tests")
            )

    member private this.itemOnPropertyChangedHandler =
        new PropertyChangedEventHandler(fun _ args -> this.itemOnPropertyChanged args) 
            
    member this.itemOnPropertyChanged (propertyChangedEventArgs:PropertyChangedEventArgs) = 
        if propertyChangedEventArgs.PropertyName = "Results"
        then this.OnPropertyChanged("Results")

        if propertyChangedEventArgs.PropertyName = "Tests"
        then this.OnPropertyChanged("Tests")

    member this.Add _ = 
        let fileOpen = new Microsoft.Win32.OpenFileDialog ();
        fileOpen.Filter <- "test suites|*.dll;*.exe"
        fileOpen.Multiselect <- false

        let result = fileOpen.ShowDialog()
        if result.GetValueOrDefault() = true
        then
            if assemblies |> Seq.filter(fun a -> a.Name = fileOpen.FileName) |> Seq.length = 0
            then
                let testAssemblyModel = new TestAssemblyModel(fileOpen.FileName) :> ITestAssemblyModel
                assemblies.Add(testAssemblyModel)

                let dict = new Dictionary<string, EventHandler>()


                let onDelete = 
                    new EventHandler (
                        fun sender args ->
                            let sender = sender :?> ITestAssemblyModel
                            let self = this :> ITestsMainModel
                            (self.Assemblies).Remove(sender) |> ignore
                            sender.DeletedFile.RemoveHandler(dict.["onDelete"])
                    )

                dict.Add("onDelete", onDelete)

                testAssemblyModel.DeletedFile.AddHandler onDelete

    member this.Run _ = 
        this |> runIt (fun model -> model.Run null)

    member this.Debug _ = 
        this |> runIt (fun model -> model.Debug null)

    member private this.GetTestItems (selector: ITestAssemblyModel -> 'b seq) =
        query { for assembly in assemblies do
                    select (selector assembly)
                } |> Seq.collect (fun a -> a)

    interface ITestsMainModel with
        member this.Assemblies
            with get () = assemblies
            and set value = 
                if value <> assemblies
                then
                    assemblies <- value
                    this.OnPropertyChanged("Assemblies")

        member this.IsRunning 
            with get () = isRunning
            and set value = 
                if value <> isRunning
                then
                    isRunning <- value
                    this.OnPropertyChanged("IsRunning")

        member this.Description
            with get () = 
                selected.Name + "\n\n" + selected.FailDetail

        member this.Selected
            with get () = selected
            and set value =
                if value <> selected
                then
                    selected <- value
                    this.OnPropertyChanged("Selected")
                    this.OnPropertyChanged("Description")

        member this.Results
            with get () = this.GetTestItems (fun (a:ITestAssemblyModel) -> a.Results :> TestResult seq) |> Seq.toArray

        member this.Tests
            with get () = this.GetTestItems (fun (a:ITestAssemblyModel) -> a.Tests :> ITestDetailModel seq) |> Seq.toArray

        member this.RunCommand
            with get () = new DelegateCommand((fun _ -> this.Run(null)), fun _ -> not this.ITestsMainModel.IsRunning) :> ICommand

        member this.DebugCommand
            with get () = new DelegateCommand((fun _ -> this.Debug(null)), fun _ -> not this.ITestsMainModel.IsRunning) :> ICommand

        member this.AddCommand
            with get () = new DelegateCommand(fun _ -> this.Add(null)) :> ICommand

    member private this.ITestsMainModel = this :> ITestsMainModel

    member this.Assemblies
        with get () = this.ITestsMainModel.Assemblies
        and set value = this.ITestsMainModel.Assemblies <- value

    member this.IsRunning
        with get () = this.ITestsMainModel.IsRunning
        and set value = this.ITestsMainModel.IsRunning <- value

    member this.Description
        with get () = this.ITestsMainModel.Description

    member this.Selected
        with get () = this.ITestsMainModel.Selected
        and set value = this.ITestsMainModel.Selected <- value

    member this.Results
        with get () = this.ITestsMainModel.Results

    member this.Tests
        with get () = this.ITestsMainModel.Tests

    member this.RunCommand
        with get () = this.ITestsMainModel.RunCommand
        
    member this.DebugCommand
        with get () = this.ITestsMainModel.DebugCommand

    member this.AddCommand
        with get () = this.ITestsMainModel.AddCommand